contracts/helpers/TransferHelper.sol

```diff
@@ -42,123 +55,344 @@ contract TransferHelper is TransferHelperInterface, TokenTransferrer {
+                totalItems += transfer.items.length;
+            }
+        }
+
+        // Declare a new array in memory with length totalItems to populate with
+        // each conduit transfer.
+        ConduitTransfer[] memory conduitTransfers = new ConduitTransfer[](
+            totalItems
+        );
+
+        // Declare an index for storing ConduitTransfers in conduitTransfers.
+        uint256 itemIndex;
+
+        // Skip overflow checks: all for loops are indexed starting at zero.
+        unchecked {
+            // Iterate over each transfer.
+            for (uint256 i = 0; i < totalTransfers; ++i) {
+                // Retrieve the transfer in question.
+                TransferHelperItemsWithRecipient calldata transfer = transfers[
+                    i
+                ];
+                TransferHelperItem[] calldata transferItems = transfer.items;
 
-            // Skip overflow checks: all for loops are indexed starting at zero.
-            unchecked {
-                // Iterate over each transfer.
-                for (uint256 i = 0; i < totalTransfers; ++i) {
-                    // Retrieve the transfer in question.
-                    TransferHelperItem calldata item = items[i];
+                // Ensure recipient is not the zero address.
+                _checkRecipientIsNotZeroAddress(transfer.recipient);
+
+                // Retrieve total number of transfers and place on stack.
+                uint256 totalItemTransfers = transferItems.length;
+
+                // Iterate over each item in the transfer to create a
+                // corresponding ConduitTransfer.
+                for (uint256 j = 0; j < totalItemTransfers; ++j) {
+                    // Retrieve the item from the transfer.
+                    TransferHelperItem calldata item = transferItems[j];
+
+                    if (item.itemType == ConduitItemType.ERC20) {
+                        // Ensure that the identifier of an ERC20 token is 0.
+                        if (item.identifier != 0) {
+                            revert InvalidERC20Identifier();
+                        }
+                    } else if (item.itemType == ConduitItemType.ERC721) {
+                        // Ensure that the amount of an ERC721 token is 1.
+                        if (item.amount != 1) {
+                            revert InvalidERC721TransferAmount();
+                        }
+                    }
+
+                    // If the item is an ERC721 token and
+                    // callERC721Receiver is true...
+                    if (item.itemType == ConduitItemType.ERC721) {
+                        // Create a boolean indicating whether
+                        // validateERC721Receiver is true and recipient is
+                        // a contract.
+                        bool callERC721Receiver = transfer
+                            .validateERC721Receiver &&
+                            transfer.recipient.code.length != 0;
```

Comment:
This code size check should probably be performed once, outside the loop, and reused here; it’ll increase overhead in cases where there aren’t actually any 721 transfers but will cut down quite a bit in cases where there are a bunch of them

Reply:
And actually can be skipped when the flag isn’t set anyway!